
# The pupose of this script is to recreate the PCAs to visualize the boards, pigs, hybrids and PCA genetic structure on a figure. The authors chosed to run PCAs with a pearl script implemented in EIGENSOFT 3.0 (Price et al. 2006). As the pearl script didn't perform well, the documentation rare and as R users we chose to perform PCAs with the R package adegenet.

## Install the required packages 

```{r}
install.packages("adegenet")
```

## Create a population information factor: 

# A previous analysis on the SNPs board data done with STRUCTURE has identified 4 genetic cluster. These are indicated in the Boar_info.txt and will be helpfull to visualize the clusters on the PCA. 
# We first create a factor will the PCA.cluster information from the Boar_info.txt file.

```{r}
pop_info <- read.table("Boar_info.txt", sep="\t", h=T, stringsAsFactors = F)
pop_info$PCA.cluster <- as.factor(pop_info$PCA.cluster)
pop_info_pca <- pop_info$PCA.cluster
View(pop_info_pca)
```

# The second PCA coutains 9 pig genotypes added to the board data set. We will visualize and copy the population information factor and simply recreate a new one with 9 more values corresponding to a 5th cluster: the pigs from different breeds. This factor will be used later.

```{r}
pop_info_pca # Visualize and copy the factor.
all_info_pca <- c("1","1","1","1","1","1","1","1","1","1","1","1","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","3","2","1","2","1","2","2","2","1","hyb1","1","1","2","4","2","2","1","hyb3","1","1","1","2","1","1","1","2","2","2","2","2","2","2","2","2","2","hyb5","hyb2", "hyb6","2","hyb4","2","4","4","4","hyb8","4","4","4","hyb7", "hyb9","4","4","4","4","5","5","5","5","5","5","5","5","5")
all_info_pca <- as.factor(all_info_pca)
View(all_info_pca)
```

## PCA with adegenet:

# We used the adegenet packages and used the codes proposed by Thibault Jombart from in the adegenet tutorial. (http://adegenet.r-forge.r-project.org/files/tutorial-basics.pdf)
# To run the following code we need to have previously created an data.frame containing the SNP genotypes wanted (see snpStats.Rmd).
# The PCA is performed using the Board data set with minor allele frequencies (MAF) inferior to 0.05 removed.
# The first step is to convert the SNP data frame into a "genind" object. The ploidy (here n=2) and the coding format (here 1) need to be specified.

```{r}
library(adegenet)
obj <- df2genind(board_filtered_0.05_genotypes, ploidy=2, ncode=1)
```

# The following performs the PCA
# The console will ask for how many axes to select. Since most of the variance is explained by the 4 first axes, we will enter "4".

```{r}
pca = dudi.pca(obj, center=TRUE, scale=FALSE)
```

# Visualize the PCA

```{r}
s.label(pca$li)
```

# This first figure is not very helpfull because of labels overlap but we can improve the figure.

```{r}
s.class(pca$li, fac=pop_info_pca, col=funky(13))
```

# A better visualization with:

```{r}
s.class(pca$li, fac=pop_info_pca,
        xax=1,yax=2,
        col=transp(funky(13),.6),
        axesel=FALSE, cstar=0, cpoint=3, grid = F)
add.scatter.eig(pca$eig[1:50],3,1,2, ratio=.3)
```

## PCA including some pig genotypes

# The SNPs data frame incluing some pig genotypes needs to be merged prior to run the following code. See snpStats.Rmd.

```{r}
obj2 <- df2genind(all_genotypes, ploidy=2, ncode=1)
pca2 = dudi.pca(obj2, center=TRUE, scale=FALSE)
s.label(pca2$li)
s.class(pca2$li, fac=all_info_pca, col=funky(13))
s.class(pca2$li, fac=all_info_pca,
        xax=1,yax=2,
        col=transp(funky(13),.6),
        axesel=FALSE, cstar=0, cpoint=3, grid = F)
add.scatter.eig(pca$eig[1:50],3,1,2, ratio=.3)
```